# Assignment - Part 1

### Here we will do the cleaning of the data during the scraping operation itself. We will remove the postal codes with no boroughs assigned and also do all the required formattings.

In [85]:
from bs4 import BeautifulSoup
import requests
import csv
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(source, 'lxml')
table = soup.find('table')
tbody = table.find('tbody')

csv_file = open('canada_ps.csv', 'w')
csv_writer = csv.writer(csv_file)
csv_writer.writerow(['PostalCode', 'Borough', 'Neighborhood'])

for tr in tbody.find_all('tr')[1:]:
    post = tr.find_all('td')[0].text.strip()
    borough = tr.find_all('td')[1].text.strip()
    neighborhood = tr.find_all('td')[2].text.strip()
    if borough != 'Not assigned':
        if (len(neighborhood.strip()) == 0):
            neighborhood = borough
        try:
            neighborhood = neighborhood.replace(" /",",")
        except:
            pass
        csv_writer.writerow([post, borough, neighborhood])
        
csv_file.close()

Now, we have the formatted data stored in the csv file. Let us now import the data as a pandas data frame

In [86]:
import pandas as pd
df = pd.read_csv('canada_ps.csv')

In [138]:
df_new = df.set_index('PostalCode')
df_display = df_new.loc[['M5G','M2H','M4B','M1J','M4G','M4M', 'M1R','M9V','M9L','M5V','M1B','M5A']]
df_display.reset_index(inplace=True)
df_display

PostalCode           Borough  \
0         M5G  Downtown Toronto   
1         M2H        North York   
2         M4B         East York   
3         M1J       Scarborough   
4         M4G         East York   
5         M4M      East Toronto   
6         M1R       Scarborough   
7         M9V         Etobicoke   
8         M9L        North York   
9         M5V  Downtown Toronto   
10        M1B       Scarborough   
11        M5A  Downtown Toronto   

                                         Neighborhood  
0                                  Central Bay Street  
1                                   Hillcrest Village  
2                     Parkview Hill, Woodbine Gardens  
3                                 Scarborough Village  
4                                             Leaside  
5                                     Studio District  
6                                   Wexford, Maryvale  
7   South Steeles, Silverstone, Humbergate, Jamest...  
8                                       Humber Summit  
9   CN Tower, King and Spadina, Railway Lands, Har...  
10                                     Malvern, Rouge  
11                          Regent Park, Harbourfront

In [88]:
df.shape

(103, 3)

The scraped data frame consists of 103 rows

# Assignment - Part 2 

Now, let us add the latitudes and longitudes using the geocoder package

Tried to use the geocoder package, but it is taking a lot of time to fetch the locations. So let us use the provided link to get the location data

In [89]:
!wget -O canada_loc.csv http://cocl.us/Geospatial_data

--2020-04-25 11:30:09--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 159.8.69.24, 159.8.72.228, 159.8.69.21
Connecting to cocl.us (cocl.us)|159.8.69.24|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2020-04-25 11:30:09--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|159.8.69.24|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-04-25 11:30:11--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 107.152.26.197
Connecting to ibm.box.com (ibm.box.com)|107.152.26.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-04-25 11:30:11--  https://ibm.box.com/public/static/9a

In [90]:
df_loc = pd.read_csv('canada_loc.csv')

In [91]:
df_loc.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [92]:
df_loc.rename(columns = {'Postal Code':'PostalCode'}, inplace = True)
df_merged = df.merge(df_loc, on='PostalCode', how='left')
df_merged.head(10)

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke                             Islington Avenue   
6        M1B       Scarborough                               Malvern, Rouge   
7        M3B        North York                                    Don Mills   
8        M4B         East York              Parkview Hill, Woodbine Gardens   
9        M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

# Assignment - Part 3

As given in the instructions for the assignment, let's first extract the Boroughs with names that contain 'Toronto'

In [93]:
df_toronto = df_merged[df_merged['Borough'].str.contains('Toronto')].reset_index(drop=True)
print(df_toronto.shape)
df_toronto.head()

(39, 5)


PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

We can see that there are 39 such entries

The geographical coordinates of Toronto is 43.6532 and -79.3832 (obtained from a Google Search). Let's create a folium map with all the neighborhoods under consideration.

First let's import the folium package

In [94]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: done

# All requested packages already installed.



In [95]:
latitude = 43.6532
longitude = -79.3832

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Now let's go ahead and use the foursquare API to fetch information about the neighborhoods and then cluster them.

#### Foursquare Credentials and Version

In [96]:
# The code was removed by Watson Studio for sharing.

Let's define a function that can help us do this easily for all neighborhoods

In [97]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [98]:
LIMIT = 100
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town,

In [99]:
print(toronto_venues.shape)
toronto_venues.head()

(1612, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Let's check how many unique venue categories are there in our dataframe

In [100]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 unique categories.


#### Now let's use one hot encoding and then group by neighbourhoods to find out the relative frequency of occurence of each venue category in a neighborhood

In [101]:
#one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

#formatting the dataframe with 'Neighborhood in the first column'
toronto_onehot.insert(0, 'Neighborhood Name', toronto_venues['Neighborhood'])

toronto_grouped = toronto_onehot.groupby('Neighborhood Name').mean().reset_index()
toronto_grouped

Neighborhood Name   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2               Business reply mail Processing CentrE  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.055556   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                      Commerce Court, Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                       Dufferin, Dovercourt Village  0.000000   
11             First Canadian Place, Underground city  0.000000   
12                           Forest Hill North & West  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East, Union Station, Toronto Islands  0.000000   
15                      High Park, The Junction South  0.000000   
16                     India Bazaar, The Beaches West  0.000000   
17          Kensington Market, Chinatown, Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                           Little Portugal, Trinity  0.000000   
20                        Moore Park, Summerhill East  0.000000   
21                                 North Toronto West  0.000000   
22                             Parkdale, Roncesvalles  0.000000   
23        Queen's Park, Ontario Provincial Government  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29                                     St. James Town  0.000000   
30                        St. James Town, Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
34                The Annex, North Midtown, Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                       The Danforth West, Riverdale  0.000000   
37           Toronto Dominion Centre, Design Exchange  0.000000   
38                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.055556      0.055556        0.111111         0.166667   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000     

Let's see what are the top 3 venues in each neighborhood

In [102]:
num_top_venues = 3

for hood in toronto_grouped['Neighborhood Name']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood Name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.05
2    Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09


----Business reply mail Processing CentrE----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2                 Spa  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.17
1    Airport Lounge  0.11
2  Airport Terminal  0.11


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.21
1  Italian Restaurant  0.06
2                Café  0.05


----Christie----
           venue  freq
0  Grocery Store  0.22
1           Café  0.17
2           Park  0.11


----Church and Wellesley----
                 venue  freq
0     Sushi Restaurant  0.08
1  Japanese Restaur

Let's create a dataframe with the top 10 most common venu categories for each neighborhood. We will first define a function, that can help us find out the top 10 venues and return it in a sorted format so that our operation becomes easier.

In [103]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [104]:
import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood Name']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood Name'] = toronto_grouped['Neighborhood Name']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood Name 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing CentrE    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Italian Restaurant              Beer Bar   
1        Breakfast Spot           Coffee Shop     Convenience Store   
2           Yoga Studio                   Spa         Garden Center   
3        Airport Lounge      Airport Terminal               Airport   
4    Italian Restaurant                  Café        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0            Restaurant                  Café                Cheese Shop   
1            Restaurant                   Bar                     Bakery   
2                Garden  Fast Food Restaurant             Farmers Market   
3       Harbor / Marina           Coffee Shop                      Plane   
4       Bubble Tea Shop          Burger Joint  Middle Eastern Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery    Seafood Restaurant         Farmers Market  
1          Intersection          Climbing Gym     Italian Restaurant  
2            Comic Shop           Pizza Place       Recording Studio  
3   Rental Car Location      Sculpture Garden          Boat or Ferry  
4   Japanese Restaurant           Salad Place         Ice Cream Shop

## Clustering

Let's go ahead and cluster the negihborhoods into 5 clusters using k-means clustering

In [110]:
from sklearn.cluster import KMeans

kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood Name', 1)

kmeans = KMeans(n_clusters=kclusters,init='k-means++', n_init=10, random_state=0).fit(toronto_grouped_clustering)

neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

#adding the clusters to the another dataframe with other relevant information

neighborhoods_venues_sorted.rename(columns = {'Neighborhood Name': 'Neighborhood'}, inplace = True)

In [111]:
toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               4                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                   Pub                  Park   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café   Japanese Restaurant   
3                  Café             Gastropub          Cocktail Bar   
4     Health Food Store                   Pub          Neighborhood   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot            Restaurant                  Café   
1              Creperie              Beer Bar        Sandwich Place   
2            Restaurant        Cosmetics Shop       Bubble Tea Shop   
3   American Restaurant    Italian Restaurant           Art Gallery   
4          Dessert Shop    Dim Sum Restaurant                 Diner   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Theater    Mexican Restaurant             Shoe Store  
1               Burger Joint         Burrito Place                   Café  
2  Middle Eastern Restaurant        Lingerie Store   Fast Food Restaurant  
3         Seafood Restaurant        Clothing Store    Moroccan Restaurant  
4             Discount Store           Yoga Studio          Deli / Bodega

Now let's visualise our clusters

In [112]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining the clusters

#### Cluster 1

In [113]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                           Regent Park, Harbourfront               0   
1         Queen's Park, Ontario Provincial Government               0   
2                            Garden District, Ryerson               0   
3                                      St. James Town               0   
5                                         Berczy Park               0   
6                                  Central Bay Street               0   
7                                            Christie               0   
8                            Richmond, Adelaide, King               0   
9                        Dufferin, Dovercourt Village               0   
10  Harbourfront East, Union Station, Toronto Islands               0   
11                           Little Portugal, Trinity               0   
12                       The Danforth West, Riverdale               0   
13           Toronto Dominion Centre, Design Exchange               0   
14       Brockton, Parkdale Village, Exhibition Place               0   
15                     India Bazaar, The Beaches West               0   
16                     Commerce Court, Victoria Hotel               0   
17                                    Studio District               0   
20                                   Davisville North               0   
22                      High Park, The Junction South               0   
23                                 North Toronto West               0   
24                The Annex, North Midtown, Yorkville               0   
25                             Parkdale, Roncesvalles               0   
26                                         Davisville               0   
27                     University of Toronto, Harbord               0   
28                                 Runnymede, Swansea               0   
30          Kensington Market, Chinatown, Grange Park               0   
31  Summerhill West, Rathnelly, South Hill, Forest...               0   
32  CN Tower, King and Spadina, Railway Lands, Har...               0   
34                                     Stn A PO Boxes               0   
35                        St. James Town, Cabbagetown               0   
36             First Canadian Place, Underground city               0   
37                               Church and Wellesley               0   
38              Business reply mail Processing CentrE               0   

   1st Most Common Venue 2nd Most Common Venue      3rd Most Common Venue  \
0            Coffee Shop                Bakery                        Pub   
1            Coffee Shop      Sushi Restaurant                      Diner   
2         Clothing Store           Coffee Shop                       Café   
3            Coffee Shop                  Café                  Gastropub   
5            Coffee Shop          Cocktail Bar         Italian Restaurant   
6            Coffee Shop    Italian Restaurant                       Café   
7          Grocery Store                  Café                       Park   
8            Coffee Shop                  Café                 Restaurant   
9                 Bakery              Pharmacy  Middle Eastern Restaurant   
10           Coffee Shop              Aquarium                       Café   
11                   Bar            Restaurant                Men's Store   
12      Greek Restaurant           Coffee Shop         Italian Restaurant   
13           Coffee Shop                 Hotel                       Café   
14                  Café        Breakfast Spot                Coffee Shop   
15  Fast Food Restaurant        Sandwich Place                  Pet Store   
16           Coffee Shop                  Café                 Restaurant   
17                  Café           Coffee Shop                  Gastropub   
20                  Park      Department Store             Breakfast Spot   
22    Mexican Restaurant                  Café            Thai Restaurant   
23        Clothing

#### Cluster 2

In [114]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19     Roselawn               1                Garden           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19      Department Store  Ethiopian Restaurant     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
19      Doner Restaurant                Dog Run

#### Cluster 3

In [115]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
29  Moore Park, Summerhill East               2            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29            Restaurant           Yoga Studio         Deli / Bodega   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 4

In [116]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
18             Lawrence Park               3                  Park   
21  Forest Hill North & West               3                  Park   
33                  Rosedale               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18              Bus Line           Swim School    Dim Sum Restaurant   
21         Jewelry Store                 Trail              Bus Line   
33            Playground                 Trail          Dance Studio   

          5th Most Common Venue 6th Most Common Venue  \
18         Ethiopian Restaurant     Electronics Store   
21             Sushi Restaurant      Department Store   
33  Eastern European Restaurant   Dumpling Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
18  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
21  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
33                   Donut Shop      Doner Restaurant               Dog Run   

   10th Most Common Venue  
18       Doner Restaurant  
21       Doner Restaurant  
33    Distribution Center

#### Cluster 5

In [117]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  The Beaches               4                 Trail     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                   Pub          Neighborhood          Dessert Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4    Dim Sum Restaurant                 Diner        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
4           Yoga Studio          Deli / Bodega

### Observations

We can see that most of the neighborhoods are very similar as 33 of the 39 neighborhoods got clustered into a single cluster. In Cluster 2,3 and 5, there is only one neighborhood and it implies that these neighborhoods are very different from all the others.

For cluster 1 and 4, we can see some underlying features that makes the members of these cluster similar.

Cluster 1 - Mainly consists of Coffee Shops, Cafes and food joints

Cluster 4 - Characterised by parks, Easter Europian Restaurants, Dumpling Restaurants and Donut Shops